particle_structs/src/ps_for.hpp

```diff
@@ -53,4 +53,21 @@ namespace pumipic {
     throw 1;
     return NULL;
   }
+
+  //This function initializes and populates the pids and offsets arrays
+  template <typename DataTypes, typename MemSpace>
+  template <typename ViewT>
+  void ParticleStructure<DataTypes, MemSpace>::getPIDs(ViewT& pids, ViewT& offsets) {
+    pids = ViewT("pids", capacity_);
+    offsets = ViewT("offsets", num_elems+1);
+    ViewT ppe("ppe", num_elems+1);
+    auto setPIDs = PS_LAMBDA(const lid_t& e, const lid_t& p, const bool& mask) {
+      if (mask) {
+        pids(p) = p;
+        Kokkos::atomic_increment(&ppe(e));
+      }
+    };
+    parallel_for(this, setPIDs, "setPIDs");
+    exclusive_scan(ppe, offsets, execution_space());
+  }
```

Comment:
@cwsmith Does this look right to you? I believe that the PIDs are wrong because they are not sorted by element, or am I wrong about that?

Reply:
At first glance, I think this will at least leave gaps in `pids` as `p` isn't necessarily contiguous as some structures have padding.  I'm thinking the following:
- parallel_for to count `ppe`
- exclusive_scan over `ppe`
- a host allocation of a `pids` array that has exactly the right amount of space using the last entry from the array created with the ex_scan (which is simply the total number of active particles)
- parallel_for with a per element counter (that is incremented atomically as done for `ppe`) that inserts particle ids into the `pids` array based on the value of the per element counter